extension/src/time_series/pipeline/lambda/executor.rs

```diff
@@ -0,0 +1,334 @@
+
+use pgx::*;
+
+use super::*;
+
+pub struct ExpressionExecutor<'e> {
+    exprs: &'e Expressions,
+    var_vals: Vec<Option<Value>>,
+}
+
+impl<'e> ExpressionExecutor<'e> {
+    pub fn new(exprs: &'e Expressions) -> Self {
+        Self {
+            var_vals: vec![None; exprs.variables.len()],
+            exprs,
+        }
+    }
+
+    pub fn reset(&mut self) {
+        for v in &mut self.var_vals {
+            *v = None
+        }
+    }
+
+    pub fn exec(&mut self, value: f64, time: i64) -> Value {
+        self.exec_expression(&self.exprs.expr, value, time)
+    }
+
+    fn exec_expression(&mut self, expr: &Expression, value: f64, time: i64) -> Value {
+        use Expression::*;
+        match expr {
+            ValueVar => Value::Double(value),
+            TimeVar => Value::Time(time),
+            DoubleConstant(f) => Value::Double(*f),
+            TimeConstant(t) => Value::Time(*t),
+            IntervalConstant(i) => Value::Interval(*i),
+
+            UserVar(i, _) => self.force_var(*i, value, time),
+
+            FunctionCall(function, args) =>
+                self.exec_function(function, args, value, time),
+
+            Unary(op, expr, ty) =>
+                self.exec_unary_op(*op, ty, expr, value, time),
+
+            Binary(op, left, right, ty) =>
+                self.exec_binary_op(*op, ty, left, right, value, time),
+
+            BuildTuple(exprs, _) =>
+                Value::Tuple(exprs.iter().map(|e| self.exec_expression(e, value, time)).collect()),
+        }
+    }
+
+    fn force_var(&mut self, i: usize, value: f64, time: i64) -> Value {
+        if let Some(value) = &self.var_vals[i] {
+            return value.clone()
+        }
+
+        let value = self.exec_expression(&self.exprs.variables[i], value, time);
+        self.var_vals[i] = Some(value.clone());
+        value
+    }
+
+    fn exec_function(
+        &mut self,
+        function: &Function,
+        args: &[Expression],
+        value: f64,
+        time: i64,
+    ) -> Value {
+        use Function::*;
+        macro_rules! unary_function {
+            ($func:ident ( )) => {
+                {
+                    let then = self.exec_expression(&args[0], value, time).float();
+                    then.$func().into()
+                }
+            };
+        }
+        macro_rules! binary_function {
```

Comment:
Unless you're planning on adding some binary functions other than `Log`, remove this macro.  (if you keep it, use it for `log`)